pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java

```diff
@@ -1947,8 +1933,34 @@ protected void triggerFlush() {
         }
     }
 
+    // must acquire semaphore before calling
+    private void maybeTriggerBatchTimerTask() {
+        if (this.batchTimerTask == null) {
+            this.batchTimerTask = cnx().ctx().executor()
+                    .schedule(catchingAndLoggingThrowables(() -> {
+                        if (log.isTraceEnabled()) {
+                            log.trace("[{}] [{}] Batching the messages from the batch container from timer thread",
+                                    topic, producerName);
+                        }
+                        // No need to prematurely cut batches if we're not ready to send them.
+                        if (getState() != State.Ready) {
+                            return;
+                        }
+                        // semaphore acquired when message was enqueued to container
+                        synchronized (ProducerImpl.this) {
+                            batchMessageAndSend();
+                        }
+                    }), conf.getBatchingMaxPublishDelayMicros(), TimeUnit.MICROSECONDS);
+        }
+    }
+
     // must acquire semaphore before enqueuing
     private void batchMessageAndSend() {
+        if (this.batchTimerTask != null) {
+            // Cancel batch timer task since we're sending the batch now. (Don't interrupt; it could be this future.)
+            this.batchTimerTask.cancel(false);
+            this.batchTimerTask = null;
+        }
```

Comment:
The previous in-flight batch task shouldn't be cancelled. That would cause problems. There can be multiple tasks queued and in flight at once. It's a valid case that the previous task might not be completed when the next batch task is created.

Reply:
My proposed design is to have either one or no tasks in-flight. If we don't cancel tasks, we'll end up with a task per batch message, which will result in smaller batches and increased CPU utilization when sending many batches.

What problems do you see cancellation causing?

This design relies on the fact that the batch task is only ever created or canceled from within a lock on `this` producer. There are no races. When we are flushing a batch, there are no pending batch messages. The next time the lock on `this` is acquired and a message is enqueued, it can correctly schedule the next batch task.

I admit that cancellation adds CPU overhead. I haven't bench marked this to compare it to the current design.

Reply:
Yes, there should only be 1 batch being packed at once, because we're reusing some of the containers

Reply:
> My proposed design is to have either one or no tasks in-flight. If we don't cancel tasks, we'll end up with a task per batch message, which will result in smaller batches and increased CPU utilization when sending many batches.

I don't see how this could work in the current solution. The chance for a data race remains: it's possible that the scheduled task has started execution and runs batchMessageAndSend concurrently with the batchMessageAndSend that results from a batch being filled up. Cancelling the previous task doesn't resolve this issue in the given example. 
batchMessageAndSend should be run within a synchronized block so it takes care of races at that level. However, in theory that this happens:
1. batchMessageAndSend gets triggered since it's full
2. a new message is added to batch
3. batchMessageAndSend gets triggered by the timer
In the above example, it's possible that the batchMessageAndSend gets sent off with just 1 message in the batch message.

> What problems do you see cancellation causing?

It doesn't prevent the data race explained in the given example. If the execution hasn't yet started, it will not run the task, but there's a chance for races after the point when the task is already executing.

Reply:
I should add that there is an issue with the earlier logic where it wouldn't have worked with a producer that had to reconnect. I just pushed up what I believe are correct changes. I will add a test later today to verify the reconnect case. The other cases are already covered in the the `BatchMessageTest`.

Reply:
@lhotari - I updated the logic to prevent early batch flushes. Please take a look.